### ISSUE TITLE
Add new API documentation for the new mining dashboard endpoints

### ISSUE THREAD

**Issue by softsimon** [mempool/mempool#1646]
### New endpoints

```
mining/pools/:interval
mining/pools/all
mining/pool/:slug/hashrate
mining/pool/:slug/blocks
mining/pool/:slug/blocks/:height
mining/pool/:slug
mining/pool/:slug/:interval
mining/hashrate/pools
mining/hashrate/pools/:interval
mining/hashrate
mining/hashrate/:interval
mining/reward-stats/:blockCount
mining/blocks/fees/:interval
mining/blocks/rewards/:interval
mining/blocks/fee-rates/:interval
mining/blocks/sizes-weights/:interval
```

# Available intervals:
```

24h
3d
1w
1m
3m
6m
1y
2y
3y

```

**Comment by hunicus** [comment]
@softsimon @nymkappa below are some items that came up as I've been working through this. 

I'm not super familiar with the reasoning that went into making the API as it is, so I'm not saying the following things are necessarily _problems_, but they seem worth pointing out in case they're not intended behavior.

- It seems that `mining/pool/:slug/:interval` returns the same data no matter which interval you specify.
- For `mining/hashrate/pools/[interval]`, 24h and 3d intervals don't seem to work.
  - Also it seems like results are ordered by hashrate, for the most part, but if this is the case, some items are slightly out of order.
- For `mining/pool/:slug/hashrate`, it seems 18 data points are returned representing 9 weeks, with 2 data points per week. What time period is `avgHashrate` measured over?
- Right now I can get a maximum of 75 days / 10 weeks of hashrate and difficulty data from `mining/hashrate/[:interval]`. Will that remain the maximum trailing time period of data available going forward?
- For `mining/blocks/fees/:interval` and `mining/blocks/rewards/:interval`, metrics are labeled as averages but aren't averages (`avg_height`, `avg_rewards`, `avg_fees`). 
  - Also intervals seem to work up to 3m, but with 6m and after they seem to get wonky.
- For `mining/blocks/fee-rates/:interval`, metrics are labeled as averages but aren't averages (`avg_height`, `avg_fee_*`, etc)...the fee figures seem like percentiles.
  - Also intervals seem to work up to 3m, but with 6m and after they seem to get wonky.
- `mining/blocks/sizes-weights/:interval` seems to have the same issue with intervals as the other blocks endpoints, as described above.
- `blocks-extras` seems to be exactly the same as `blocks`, except it returns 15 blocks instead of 10 (?).
- `block/:hash` shows an `extras` property for newer blocks but not older blocks...do we know the block at which this switch happens?
- It seems inconsistent that `mining/pools/` throws an error instead of returning data for all mining pools. Most other endpoints don't have an "all" parameter...they just offer all available data by default if an optional parameter isn't specified.

I'm going to make a draft PR soon for you guys to review what I have so far.

**Comment by nymkappa** [comment]
> It seems that mining/pool/:slug/:interval returns the same data no matter which interval you specify.
Correct, that's legacy and I forgot to update it. You can ignore the `:interval` part.

> For mining/pool/:slug/hashrate, what time period is avgHashrate measured over?
`avgHashrate` is the **estimated hashrate using data over a week**. To compute this value:
* 1/ Count how many blocks were found by the network between two Mondays at midnight UTC time
* 2/ Count how many blocks were found by the the mining pool between two Mondays at midnight UTC time
* 3/ Get the estimated hashrate from Bitcoin Core using `getnetworkhashps(network_block_count_for_a_week, last_block_height_before_monday_midnight)`
* Using 1/ and 2/ and 3/, we can therefore estimate the mining pool hashrate for that week

> For mining/hashrate/[:interval], I can only go back ~2 months on my local Mempool instance, but mempool.space goes back 3 years. Is this trailing time period something that will vary by instance?

This varies according to your mempool backend config. Specifically, if you want the full hashrate history of Bitcoin network, you need to set `config.MEMPOOL.INDEXING_BLOCKS_AMOUNT: -1` in `mempool-config.json` backend config file.
Although it's possible, indexing all blocks is an expensive operation, so it's not really recommended to do so on low spec hardware. Any decent laptop with a modern CPU can index the whole history in less than 24h.

> Maybe similar: block/:hash can show an extras property for newer blocks but not older blocks on a local instance.
If you often switch branches on your instances while working, there may be some instance where your database schema is not matching the database migration script.
If you do no switch branches, then there is maybe a bug in the database migration script. It's hard to tell.
Anyway, the API must in theory returns an `extras` fields for all blocks, otherwise the frontend javascript will crash. Feel free to truncate your `blocks` table in your db and re-index to confirm this behavior.

> For mining/blocks/fees/:interval and mining/blocks/rewards/:interval, metrics are labeled as averages but aren't averages (avg_height, avg_rewards, avg_fees).
They are average for `:interval` >= `1w`. They are exact value for `1d` and `3d` timespan. For the `1w` timespan, it may or may not be average depending on how fast blocks are found around the specified timestamp.

> blocks-extras seems to be exactly the same as blocks, except it returns 15 blocks instead of 10 (?). EDIT: I guess it takes a block height instead of a block hash.

The `/blocks` API is consumed when the mining dashboard is disabled in the frontend. I think this API can be dropped actually and the `blocks-extras` could maybe because the de-factor new `/blocks` API. I will take a look at this.

> It seems inconsistent that mining/pools/ throws an error instead of returning data for all mining pools. Most other endpoints don't have an "all" parameter...they just offer all available data by default if an optional parameter isn't specified.

Good point. That's the first API endpoint I wrote and other endpoints are less strict. I will fix. You can consider that not sending any `:interval` will fallback to `all` by default.

**Comment by hunicus** [comment]
Thank you @nymkappa for taking the time to write all of that out!! Very, very helpful. 

I think you covered everything...I'll make revisions and go through everything again just in case anything else comes up.
